pkg/services/meta/meta.go

```diff
@@ -0,0 +1,213 @@
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/nspcc-dev/neo-go/pkg/core/block"
+	"github.com/nspcc-dev/neo-go/pkg/core/state"
+	"github.com/nspcc-dev/neo-go/pkg/rpcclient"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	"go.uber.org/zap"
+	"golang.org/x/exp/maps"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	// raw storage prefixes.
+
+	// rootKey is the key for the last known state root in KV data base
+	// associated with MPT.
+	rootKey = 0x00
+)
+
+// ContainerLister is a source of actual containers current node belongs to.
+type ContainerLister interface {
+	// List returns node's containers that support chain-based meta data and
+	// any error that does not allow listing.
+	List() (map[cid.ID]struct{}, error)
+}
+
+// Meta handles object meta information received from FS chain and object
+// storages. Chain information is stored in Merkle-Patricia Tries. Full objects
+// index is built and stored as a simple KV storage.
+type Meta struct {
+	l        *zap.Logger
+	rootPath string
+	netmapH  util.Uint160
+	cnrH     util.Uint160
+	cLister  ContainerLister
+
+	m        sync.RWMutex
+	storages map[cid.ID]*containerStorage
+
+	endpoints   []string
+	timeout     time.Duration
+	magicNumber uint32
+	cliCtx      context.Context // for client context only, as it is required by the lib
+	ws          *rpcclient.WSClient
+	bCh         chan *block.Header
+	objEv       chan *state.ContainedNotificationEvent
+	cnrDelEv    chan *state.ContainedNotificationEvent
+	cnrPutEv    chan *state.ContainedNotificationEvent
+	epochEv     chan *state.ContainedNotificationEvent
+
+	objNotificationBuff chan *state.ContainedNotificationEvent
+}
+
+const objsBufferSize = 1024
+
+// New makes [Meta].
+func New(l *zap.Logger, cLister ContainerLister, timeout time.Duration, endpoints []string, containerH, nmH util.Uint160, rootPath string) (*Meta, error) {
+	storagesFS, err := os.ReadDir(rootPath)
+	if err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("read existing container storages: %w", err)
+	}
+	storagesRead := make(map[cid.ID]*containerStorage)
+	for _, s := range storagesFS {
+		sName := s.Name()
+		cID, err := cid.DecodeString(sName)
+		if err != nil {
+			l.Warn("skip unknown container storage entity", zap.String("name", sName), zap.Error(err))
+			continue
+		}
+
+		storagesRead[cID], err = storageForContainer(rootPath, cID)
+		if err != nil {
+			l.Warn("skip container storage that cannot be read", zap.String("name", sName), zap.Error(err))
+			continue
+		}
+	}
+
+	cnrsNetwork, err := cLister.List()
+	if err != nil {
+		return nil, fmt.Errorf("listing node's containers: %w", err)
+	}
+	for cID := range storagesRead {
+		if _, ok := cnrsNetwork[cID]; !ok {
+			err = storagesRead[cID].drop()
+			if err != nil {
+				l.Warn("could not drop container storage", zap.Stringer("cID", cID), zap.Error(err))
+			}
+
+			delete(storagesRead, cID)
+		}
+	}
+
+	storages := storagesRead
```

Comment:
looks like a simple renaming, no need imo

Reply:
it is. prefer not to change `*Read` var